File: algorithm/challenges/15.queue/main.clj
(require '[clojure.test :as test])

(defprotocol IQueue
  (add-el [this element])
  (remove-el [this])
  (peek-el [this]))

(defrecord Queue [q]
  IQueue
  (add-el [_ el]
    (swap! q conj el))
  (remove-el [_]
    (swap! q next))
  (peek-el [_]
    (first @q)))

(defn new-queue []
  (->Queue (atom (lazy-seq))))

(test/deftest queue-test
  (test/testing "Test queue functions"
    (let [q (new-queue)]
      (test/is (= nil (peek-el q)))
      (test/is (= (seq []) (remove-el q)))
      (test/is (= (seq [:a]) (add-el q :a)))
      (test/is (= (seq [:b :a]) (add-el q :b)))
      (test/is (= :b (peek-el q)))
      (test/is (= (seq [:a]) (remove-el q)))
      (test/is (= (seq []) (remove-el q))))))

(test/run-test queue-test)

File: algorithm/challenges/15.queue/main.js
class Queue {
  constructor() {
    this.data = [];
  }

  add(record) {
    this.data.unshift(record);
  }

  remove() {
    return this.data.pop();
  }

  get peek() {
    return this.data[this.data.length - 1];
  }
}

// Testing

const test = require('node:test');
const assert = require('node:assert');

test('Queue class', () => {
  const q = new Queue();
  q.add(1);
  q.add(2);

  assert.strictEqual(q.peek, 1);
  assert.strictEqual(q.peek, 1);
  assert.strictEqual(q.remove(), 1);
  assert.strictEqual(q.remove(), 2);
});
